src/backends/libtorch/torch_backend.cc

```diff
@@ -0,0 +1,608 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "src/backends/libtorch/torch_backend.h"
+
+#include <NvInfer.h>
+#include <stdint.h>
+#include "cuda/include/cuda_runtime_api.h"
+#include "src/core/constants.h"
+#include "src/core/logging.h"
+#include "src/core/model_config.h"
+#include "src/core/model_config_utils.h"
+#include "src/core/provider.h"
+#include "src/core/server_status.h"
+
+#include <ATen/Functions.h>
+
+namespace nvidia { namespace inferenceserver {
+
+namespace {
+
+DataType
+ConvertFromDLDataType(DLDataTypeCode dtype)
+{
+  switch (dtype) {
+    case DLDataType::kDLUInt:
+      return TYPE_UINT32;
+    case DLDataType::kDLInt:
+      return TYPE_INT32;
+    case DLDataType::kDLFloat:
+      return TYPE_FP32;
+    default:
+      break;
+  }
+
+  return TYPE_INVALID;
+}
+
+DLDataType
+ConvertToDLDataType(DataType dtype)
+{
+  switch (dtype) {
+    case TYPE_UINT32:
+      return DLDataType::kDLUInt;
+    case TYPE_INT32:
+      return DLDataType::kDLInt;
+    case TYPE_FP32:
+      return DLDataType::kDLFloat;
+    default:
+      break;
+  }
+
+  return DLDataType::Invalid;
+}
+
+}  // namespace
+
+LibTorchBackend::Context::Context(
+    const std::string& name, const int gpu_device, const int max_batch_size)
+    : name_(name), gpu_device_(gpu_device), max_batch_size_(max_batch_size)
+{
+}
+
+LibTorchBackend::Context::Context(Context&& o)
```

Comment:
This move constructor is a legacy thing that we want to get rid of in new backends. Instead make contexts_ a unique_ptr vector. See how plan_backend handles it.